### ISSUE TITLE
Error while loading shared libraries on AppImage run

### ISSUE THREAD

**Issue by akontsevich** [linuxdeploy/linuxdeploy#107]
Was lost with too many `linuxdeploy` repos on github. Hope have found the right one now :wink: So need to use `linuxdeploy`, not `./appimagetool-x86_64.AppImage`?

Hi @TheAssassin and @probonopd. Have error with either `linuxdeploy` or `./appimagetool-x86_64.AppImage` if I create AppImage in docker.

As You may remember I was able to build and run AppImage successfully under CentOS 7.x based docker image here: https://github.com/AppImage/AppImageKit/issues/998#issuecomment-543368166

However when we migrate to Ubuntu 18.04 and added `libgdal` and `boost` dependencies to the app some problems appeared with shared libraries on AppImage run:
```
error while loading shared libraries: <lib_name>: cannot open shared object file: 
No such file or directory
```

And this happens with tens or hundreds shared libraries so I can't solve the issue copying them manually to app lib dir like in the link [above](https://github.com/AppImage/AppImageKit/issues/998#issuecomment-543368166). So copying libraries like this:
```
# copy some missed libs to the bundle
libs=(libssl.so.1 libcrypto.so.1 libicui18n.so libicuuc.so libicudata.so \
     libgdal.so libarmadillo.so libproj.so libjson-c.so libqhull.so libwebp.so \
     libpoppler.so libepsilon.so libkmlbase.so libkmldom.so libkmlengine.so \
     libnetcdf.so libhdf5_serial.so libmfhdfalt.so libdfalt.so libogdi.so \
     libgeotiff.so libdapclient.so libdap.so libpcre.s)
for lib in "${libs[@]}"
do
    cp $(ldconfig -p | grep ${lib} | head -n1 | tr ' ' '\n' | grep /) ./$APP.AppDir/Qt/libs/    
done
```
does not work any more as there are too much libraries and previous 80 MB AppImage which worked everywhere grows now in more than 2 times and does not work. I saw `./linuxdeploy-x86_64.AppImage` output and it strips these libraries from final AppImage for some reason - see libs list above after adding `gdal` and `boost` (previously it was necessary to deploy manually only the following `libssl.so.1, libcrypto.so.1, libicui18n.so, libicuuc.so, libicudata.so`).

See `./linuxdeploy-x86_64.AppImage` output: [appimage_bug.txt](https://github.com/probonopd/linuxdeployqt/files/3903099/appimage_bug.txt)

After `linuxdeploy` I've tried Your tool - `./appimagetool-x86_64.AppImage` and got almost same results: gdal and other libraries are not included in docker build so AppImage does not work. Outside docker - all builds and works fine, however have size: ~ 320 MB with `appimagetool` tool (~377MB with `linuxdeploy`).

https://github.com/probonopd/linuxdeployqt does not work at all with modern distros like Ubuntu 18.04.

Any ideas? What the problem could be, how to solve? Thanks!

**Comment by TheAssassin** [comment]
Please note my follow-up [originally posted in the original issue](https://github.com/AppImage/AppImageKit/issues/1012#issuecomment-559897074). Short excerpt:

> 1. It seems that you just ran appimagetool on an AppDir you previously had run linuxdeploy on.
> 2. You seem to have expected appimagetool to do things it doesn't do: appimagetool just "converts" > an AppDir into an AppImage, it doesn't bundle additional files. It never has, it never will.
> 3. It does seem you have not understood the relation between the tools linuxdeploy, linuxdeployqt and appimagetool.
> 
> This is what I understood: You are unhappy because linuxdeploy has not bundled libraries your app requires (through `dlopen` or something), which means that when you run the resulting AppImage, your app shows error messages.
> 
> If that's correct so far, your issue is related to the *bundling of libraries*, not the creation of an AppImage.
> 
> linuxdeploy[qt] both use appimagetool to make AppImages, but only for that. Everything else is not > related to appimagetool. And your problem isn't that you can't make an AppImage, it's that your AppImage doesn't contain some files.
> 
> Therefore, your issue should also be titled, "AppImage lacks libraries dlopened by app" or so.

Are my assumptions correct so far? If not, please correct them.

I'm going to study your `.txt` document in a few minutes. Please be a little patient, you're not forgotten.

**Comment by TheAssassin** [comment]
The linuxdeploy log seems alright. Can you share your AppImage? For what exact libraries (please provide a list) is that message shown?

**Comment by akontsevich** [comment]
> Please be a little patient, you're not forgotten.

Thanks much! You always were very responsive, problem I was [lost in repos](https://github.com/AppImage/AppImageKit/issues/1012#issuecomment-559898009) and started from wrong one - [probonopd](https://github.com/probonopd/linuxdeployqt/issues/405).

> The linuxdeploy log seems alright, aside from the missing desktop file.

Desktop file is in `./$APP.AppDir/myapp.desktop`. Need to specify it in `--desktop-file=` switch explicitly?

>Can you share your AppImage? 

Sorry - can't - commercial. Based on https://github.com/mavlink/qgroundcontrol + `gdal` and `boost` dependencies now as I mentioned.

>For what exact libraries (please provide a list) is that message shown?

Show You `ldd` output for the binary? Can do for openSUSE variant (in Ubuntu docker they will slightly differ). Or go inside docker and run from there?

**Comment by akontsevich** [comment]
>This is what I understood: You are unhappy because linuxdeploy has not bundled libraries your app requires (through dlopen or something), which means that when you run the resulting AppImage, your app shows error messages.

Not exactly.

>Are my assumptions correct so far? If not, please correct them.

linuxdeploy bundled all the necessary libraries if run outside docker (openSUSE) - AppImageSize ~370 MB - works, and strips much if inside docker (Ubuntu 18.04) ~ 90 MB - shared libraries errors.

**Comment by TheAssassin** [comment]
This is pretty strange, there shouldn't be a *very* large difference, a slight difference (couple MBs) is to be expected, though.

Sharing the `ldd` output is a great idea. Please post any error message you find, too. You can strip private information, but please provide the name of the library that can't be found.

**Comment by TheAssassin** [comment]
> Desktop file is in ./$APP.AppDir/myapp.desktop. Need to specify it in --desktop-file= switch explicitly?

Your desktop file should lie in `AppDir/usr/share/applications/`. It later would be deployed to the AppDir root directory. linuxdeploy is more explicit about paths than other tools, it requires you to put everything in the respective directories in `AppDir/usr` instead of trying to guesstimate.
You can specify it with `--desktop-file`, it'll be put in the right place automatically for you.

**Comment by akontsevich** [comment]
Thanks, using now `--desktop-file` as all files not in `AppDir/usr`.

>Sharing the ldd output is a great idea. 

- `ldd` inside docker (Ubuntu): [ldd_ubuntu.txt](https://github.com/linuxdeploy/linuxdeploy/files/3906661/ldd_ubuntu.txt)
- `ldd` outside docker (openSUSE): [ldd_suse.txt](https://github.com/linuxdeploy/linuxdeploy/files/3906648/ldd_suse.txt)

>Please post any error message you find, too. 

No errors for now after I'd added `--desktop-file` and `-i` switches. 

>You can strip private information, but please provide the name of the library that can't be found.

I have mentioned the list of them [above](https://github.com/linuxdeploy/linuxdeploy/issues/107#issue-530510301). It is not complete however as I was forced to discover and add them manually as we talked (as on start it mentions only one lib error and need to rebuild to know the next one missed :wink:):      
```
     libgdal.so libarmadillo.so libproj.so libjson-c.so libqhull.so libwebp.so \
     libpoppler.so libepsilon.so libkmlbase.so libkmldom.so libkmlengine.so \
     libnetcdf.so libhdf5_serial.so libmfhdfalt.so libdfalt.so libogdi.so \
     libgeotiff.so libdapclient.so libdap.so libpcre.so etc etc etc
```
Think You can see all the necessary libraries in the `ldd` outputs and 2 AppImage build logs below:

- Inside docker (Ubuntu): [AppImage_in_docker.log](https://github.com/linuxdeploy/linuxdeploy/files/3906644/AppImage_in_docker.log) - AppImage size ~155 MB - shared libs errors
- outside docker (openSUSE): [AppImage_build.log](https://github.com/linuxdeploy/linuxdeploy/files/3906649/AppImage_build.log) - AppImage size ~370 MB - works fine

**Comment by TheAssassin** [comment]
The `ldd` differences are normal, every distro link their libraries differently.

I don't know what causes the size difference, I cannot guess that from the few information. You'd have to compare yourself. But a few MB more or less are acceptable.

The cause of your issues is nearly guaranteed not to be Docker. It's maybe missing dependencies or something in the Ubuntu container, but that's like using a VM of Ubuntu, I assume that'll cause problems, too. 

What makes me a little worried are the lines like:

```
WARNING: Not calling strip on binary ./MyApp.AppDir/usr/lib/libQt5Charts.so.5: rpath starts with $ 
WARNING: Not calling strip on binary ./MyApp.AppDir/usr/lib/libQt5Concurrent.so.5: rpath starts with $ 
...
```

Can you please check what's set as rpath already? You can use `patchelf --print-rpath` for that.

```
Output directory: ~/MyApp/deploy
sending incremental file list
MyApp
...
```

This looks like you mixed some rsync call into the "_build" log. Can you please fix that and provide a clean log for comparison? It also explains why the file is longer.

The non-Docker build log also deploys QML files, whereas the Docker one doesn't.

Overall, it seems like you didn't make clean builds on both platforms, but copied around your project and that unfortunately renders your logs pretty useless for debugging, as one can't tell the difference between the builds on both platforms.

**Comment by akontsevich** [comment]
> The `ldd` differences are normal, every distro link their libraries differently.

Ok. However think appimage tool should track these libraries dependencies automatically and put them into the bundle.
 
> I don't know what causes the size difference, I cannot guess that from the few information. You'd have to compare yourself. But a few MB more or less are acceptable.

How to get and list generated AppImage content to send to You?

> The cause of your issues is nearly guaranteed not to be Docker. It's maybe missing dependencies or something in the Ubuntu container, but that's like using a VM of Ubuntu, I assume that'll cause problems, too.

Think above is the main problem. But how to determine what is missed in the container?

> What makes me a little worried are the lines like:
> 
> ```
> WARNING: Not calling strip on binary ./MyApp.AppDir/usr/lib/libQt5Charts.so.5: rpath starts with $ 
> WARNING: Not calling strip on binary ./MyApp.AppDir/usr/lib/libQt5Concurrent.so.5: rpath starts with $ 

It is not a problem at all as we deploy Qt libraries on app build with Qt deployments mechanisms to AppDir/Qt/libs and appimage tool put them all together into the bundle and we run app by following command as You remember to catch-up the libraries on app startup: https://github.com/mavlink/qgroundcontrol/blob/master/deploy/qgroundcontrol-start.sh#L3


> Can you please check what's set as rpath already? You can use `patchelf --print-rpath` for that.

Run this in docker or outside?

```
> patchelf --print-rpath
patchelf: missing filename
```
For what files to run it?

> 
> ```
> Output directory: ~/MyApp/deploy
> sending incremental file list
> MyApp
> ...
> ```
> 
> This looks like you mixed some rsync call into the "_build" log. Can you please fix that and provide a clean log for comparison? It also explains why the file is longer.

I do not know what for this `rsync` is used: https://github.com/mavlink/qgroundcontrol/blob/master/deploy/create_linux_appimage.sh#L67 - what to do with this call? Remove?

> The non-Docker build log also deploys QML files, whereas the Docker one doesn't.

How do You see that? All qml files should be there as well.

> Overall, it seems like you didn't make clean builds on both platforms, but copied around your project and that unfortunately renders your logs pretty useless for debugging, as one can't tell the difference between the builds on both platforms.

For all the binaries clean build are performed, however after this `linuxdeploy` called of course to pack all of these.

**Comment by akontsevich** [comment]
Yes, Your are right: most size difference only in unnecessary `/qml` and `plugin/` files deployed.

We solved the issue **_for Ubuntu_** by copying all the libraries we got from `ldd` output:
```
# copy all libs to the bundle
cp $(ldd $QGC_RELEASE_DIR/MyApp | awk '$2 == "=>" { print $3 }') $APP.AppDir/Qt/libs
```
It works as we are running not the `MyApp` executable binary directly but wrapped it into `MyApp.sh` script like this:
https://github.com/mavlink/qgroundcontrol/blob/master/deploy/qgroundcontrol-start.sh#L3

Now AppImage is only 190 MB size. Strange why `linuxdeploy` can't track this automatically as I specify this switch `-e ./$APP.AppDir/MyApp` on app image creation?

However AppImage run still fails on **_other distros_**:
```
/docker> ./MyApp.AppImage
mkdir: /tmp/.mount_MyAppnquBVq/Qt/libs/libc.so.6: version `GLIBC_2.30' not found (required by /lib64/libselinux.so.1)
*** stack smashing detected ***: <unknown> terminated
/tmp/.mount_MyAppnquBVq/AppRun: line 11:  6558 Aborted                 (core dumped) "${HERE}/MyApp" "$@"
```
Which is strange as I have GLIBC_2.30 in openSUSE. Think not all libraries should be bundled manually.

**Comment by TheAssassin** [comment]
> Ok. However think appimage tool should track these libraries dependencies automatically and put them into the bundle.

That doesn't make any sense. linuxdeploy just deploys whatever `ldd` says the binary requires. It does exclude a few libraries following this list: https://github.com/AppImage/pkg2appimage/blob/master/excludelist
You cannot randomly copy files into your AppDir. As expected, your application crashed on another distribution.

> Strange why linuxdeploy can't track this automatically as I specify this switch -e ./$APP.AppDir/MyApp on app image creation?

You put your files into non-standard locations. linuxdeploy requires an FHS like structure. Your executable should go into `usr/bin`, your libraries need to go into `usr/lib`, etc. I start to see the greater picture why your deployment fails. We use the FHS to avoid hackery and workarounds, it's a well known concept and has proven to work for many, many apps so far.

You should probably not look *too* much at what this weird deployment script of the other app does. It's one huge workaround, and not very well written. Please share your entire script, I'll have a look then. This whole "copy this there, that there, rsync here etc." is not a good idea.

**Comment by akontsevich** [comment]
Yeah, I see You are right. Need to follow FHS structure. I just wanted to alter existing QGroundcontrol  [create_linux_appimage.sh](https://github.com/mavlink/qgroundcontrol/blob/master/deploy/create_linux_appimage.sh) AppImage build script with minimal changes to make it work. QGroundcontrol have many similar hacks for some reason.

Seems now need to rework our build -> deploy -> app_image build process to be sure it will run correctly. Will ask also if we can share our build script which is mostly like QGC one.

**Comment by TheAssassin** [comment]
Is your app *that* complex? It doesn't seem so. You should rather follow [our examples](https://docs.appimage.org/packaging-guide/from-source/native-binaries.html#examples), they already do everything the right way. As long as your qmake/CMake supports `make install`, you're pretty much done.

**Comment by akontsevich** [comment]
I follow Your examples, only not preserving FHS as you have `-e` switch. Problem left from QGC  - we deploy Qt libs from Qt SDK to the bundle to have exact Qt version, not that which comes from the distro we build in. In [your example](https://docs.appimage.org/packaging-guide/from-source/native-binaries.html#examples) if You add

`make deploy` step after following lines:

```
qmake "$REPO_ROOT"

# build project and install files into AppDir
make -j$(nproc)
make install INSTALL_ROOT=AppDir
```
it wil do this. So this hack will be necessary: https://github.com/mavlink/qgroundcontrol/blob/master/deploy/qgroundcontrol-start.sh#L3 -  we have same run script.

**Comment by TheAssassin** [comment]
Looks like you misunderstand me: `-e` etc. *copy* the files into an FHS-like structure. So, your binary will end up in the AppDir more than once. And the one in `AppDir/MyApp` is not touched by linuxdeploy. This likely causes your issues.

**Comment by akontsevich** [comment]
Our appimage build script attached - it accepts release build script as an argument:

[create_linux_appimage.sh.txt](https://github.com/linuxdeploy/linuxdeploy/files/3918937/create_linux_appimage.sh.txt)

MyApp is a QGC fork. However think this won't help much now as we come to conclusion: need to rework it to follow FHS structure.
